### ISSUE TITLE
Packaging

### ISSUE THREAD

**Issue by mpmdean** [shenmidelin/edrixs#4]
How best to distribute EDRIXS. Likely a conda / pip package. Can we compile the fortran as well as the python?

**Comment by stuartcampbell** [comment]
Yes, shouldn't be an issue to compile the fortran.  @CJ-Wright and @mrakitin do you see any issues in us trying to create a package for this ?

**Comment by CJ-Wright** [comment]
This shouldn't be an issue.

**Comment by mpmdean** [comment]
Great

**Comment by mrakitin** [comment]
I noticed the project is not structured as a proper Python package, i.e. there is no `setup.py`. Is it possible to restructure it first?

**Comment by stuartcampbell** [comment]
@mrakitin I think that is actually the first thing we need to do.

**Comment by mpmdean** [comment]
@stuartcampbell 
Do you think there might be bandwidth for someone to do this?

 @mrakitin I think @shenmidelin and I have enough knowledge to put a setup.py and I have done this previously, but I would guess it's probably better than one person does both the setup.py and the conda packaging?

**Comment by stuartcampbell** [comment]
Once the setup.py is done then the conda packing is really trivial.  I spoke to @mrakitin about it and he seemed willing to help - but I know he is super busy at the moment as well.  I will let him comment on if / when he could help.  

But in general, I think that someone should be able to help - it depends on how demanding beamlines are to how much time we have 😄

**Comment by mrakitin** [comment]
@mpmdean, thanks, that would be useful. I don't think it's necessary for one person to do both types of packaging. We package a bunch of external pip packages into conda ones, so it should not be a problem. I suggest we use https://github.com/NSLS-II/scientific-python-cookiecutter to create a template with the preconfigured structure for this project.

One question: what is the relationship of the Fortran code and the Python code? Is it a Fortran extension called from Python or is it executables called by subprocess?

**Comment by shenmidelin** [comment]
@mrakitin, Thanks ! Currently, Fortran code are standalone executables, which can be launched manually, for example, "mpirun -np 4 ed.x", or called from Python using subprocess.

**Comment by mpmdean** [comment]
Hi @shenmidelin, @mrakitin and @stuartcampbell 

I think it would probably make sense to sit down together. 

The Fortan/python relationship is more something that we should think about as edrixs is developed than a set-in-stone condition that we need to follow.

Using subprocess is a reasonable way of doing it, but I'm not knowledgeable to know if it's the ideal proceedure. From a user perspective, it would be nice to interact purely with python.

**Comment by mrakitin** [comment]
@mpmdean, yes, it makes sense to meet and discuss in details. I haven't dealt with the Fortran extensions before, but would like to learn more about it. If it's eventually a `*.so` library, it would be great. Also, in Python those parallel jobs can be parallelized with mpi4py, however looking at a more modern approach as [dask](https://dask.org/) may make sense. I think inviting @danielballan to the meeting will be beneficial as well. Please schedule a meeting.

**Comment by shenmidelin** [comment]
@mpmdean, @mrakitin, Make fortran interface with python (as *.so library) has been included in my plan of future development.
